proposals/low-level-struct-improvements.md

```diff
@@ -112,7 +112,23 @@ ref struct ReadOnlyExample
 }
 ```
 
-A `readonly ref struct` will require that `ref` fields are marked as `readonly ref`. There is no requirement that they are marked as `readonly ref readonly`. This does allow a `readonly struct` to have indirect mutations via such a field but that is no different than a `readonly` field that pointed to a reference type today.
+A `readonly` on a `ref struct` will instance will force `ref` fields to be `readonly ref readonly`. This also means a `readonly ref struct` must mark `ref` fields as `readonly ref readonly`. This maintains the invariant that a `ref struct` which is `readonly` cannot store ref-like state.
+
+```c#
+ref struct S
+{
+    ref int Field;
+
+    void Uses(in S s1, ref S s2)
+    {
+        int local = default;
+        s1.Field = 42;          // Error: can't assign ref readonly value
```

Comment:
>s1.Field = 42; [](http://example.com/codeflow?start=8&length=14)

Does the `ref readonly`-ness apply through properties as well? For instance:
```csharp
ref int Property => ref Field;
...
s1.Property = 42;
```

Reply:
For properties or methods it depends on the type of `this`. In a `readonly` property or method `this` is `ref readonly` hence all `ref` fields will be `ref readonly`.